### PR TITLE
circuit project

### PR DESCRIPTION
**Pull request by acoatoitgs**
Authors:
- Sofia Maragò @LaMatematica14
- Andrea Coato @acoatoitgs
- Jonathan Braun @Jorobraun
- Zeno Orglmeister @Meteoritini

### CODE REVIEW

File: circuit/.gitignore [hunk 1]
```diff
@@ -0,0 +1,24 @@
+*/build/
```
Comment (veluca93) [inline circuit/.gitignore]: Why this and not */exec/?

File: circuit/README.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+└── README.md                  # this file
+```
+
+---
+
+## How to run
+
+1. Create a Python environment with the following (minimum) dependencies:
+
+   * Python 3.10+ (repo shows Python 3.13 bytecode but the code is standard Python 3)
```
Comment (veluca93) [inline circuit/README.md]: bytecode?

File: circuit/converter/convert_abc.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+
+def convert_abc_format():
+    global node_count, first_output_id, input_nodes, toposorted_nodes, final_nodes, node_remap
+
+    fin = open(TRAINED_NETWORK_16GATES_FILE, "rb")
+
+    fout = open(
+        TRAINED_NETWORK_2GATES_FILE, "wb"
+    )  # binary, because encode writes binary bytes
```
Comment (veluca93) [inline circuit/converter/convert_abc.py]: "binary bytes"?

File: circuit/converter/converter.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+def replace_gates():
+    """
+    Replaces multi-input gates with NAND/NOT combinations exactly mirroring the C++ switch cases.
+    """
+    global next_free_node
+
+    for i in range(len(toposorted_nodes)):
+        new_nodes = []
+
+        __input_a = toposorted_nodes[i].link_a
```
Comment (veluca93) [inline circuit/converter/converter.py:101]: Let's get rid of `__` in front of variable names, please.

File: circuit/evaluator/evaluator.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            i += 1
+
+        # Return a float (g/cnt). If cnt==0 this will raise ZeroDivisionError (reflects unexpected inputs).
+        return g / cnt
+
+
+def make_test(net: AndNot_network, infile):
+    """
+    Read one sample from text `infile`, feed into `net`, compute network, and return net.guess(correct).
+    Mirrors the C++ make_test: read one line of bits, then one line with correct label.
```
Comment (veluca93) [inline circuit/evaluator/evaluator.py:126]: Can you get rid of the various "replicated from C++" comments? If this is the code being submitted, and no C++ is submitted, they have limited utility.

File: circuit/generator/generator.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                ]
+                first_layer_node += prev_size
+
+            # write edges left->right
+            cnt = 785
+            for l, r in zip(nl, nr):
+                f.write(f"{l} {r} {cnt}\n")
+                cnt += 1
+
+        # optional footer
```
Comment (veluca93) [inline circuit/generator/generator.py]: What is this?

File: circuit/network/network.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+        [0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.2, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1]
+    )
+    + 1.0
+)
+batch_fitting_function = jax.jit(
+    jax.vmap(lambda a: jnp.multiply(a, FITTING), in_axes=(0))
+)
+
+
+DEBUG = 0
```
Comment (veluca93) [inline circuit/network/network.py]: This should be True/False

File: circuit/network/network.py [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+        l[id] = cl
+
+    for i in range(len(layered_id)):
+        layered_id[i] = jnp.array(layered_id[i])
+
+    return layered_id, LEFT, RIGHT, PROB_ID
+
+
+def read_values(file):
+    # Autosaving
```
Comment (veluca93) [inline circuit/network/network.py]: More "Caching"

**Review by veluca93 — COMMENTED** [review]
General comment on the commit: please write a better description, and set up Co-Authored-By properly.

**Comment by veluca93** [comment]
Please squash the commits and resolve the remaining comments.
